autodiff/forward/real/real.hpp

```diff
@@ -0,0 +1,835 @@
+//                  _  _
+//  _   _|_ _  _|o_|__|_
+// (_||_||_(_)(_|| |  |
+//
+// automatic differentiation made easier in C++
+// https://github.com/autodiff/autodiff
+//
+// Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+// Copyright (c) 2018-2020 Allan Leal
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and `associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+// C++ includes
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <tuple>
+#include <utility>
+
+// autodiff includes
+#include <autodiff/common/binomialcoefficient.hpp>
+#include <autodiff/common/numbertraits.hpp>
+#include <autodiff/common/meta.hpp>
+
+namespace autodiff {
+namespace detail {
+
+/// The type used to represent a real number that supports up to *N*-th order derivative calculation.
+template<size_t N, typename T>
```

Comment:
I create this issue #86 to make sure we have this eventually. This needs to be done more carefully, because we need to permit also custom/user defined numbers to be used with `Real` and `Dual` (e.g., high-precision floating point number types).